File: Python3/102. Binary Tree Level Order Traversal.py
# 28ms, 97.38%; 13MB, 100.00%
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def levelOrder(self, root: TreeNode) -> list:
        res, q = [], []
        if root: q.append(root)
        while q:
            res.append([])
            new_q = []
            for node in q:
                res[-1].append(node.val)
                if node.left: new_q.append(node.left)
                if node.right: new_q.append(node.right)
            q = new_q
        return res
    

File: 1 - 200/102 Binary Tree Level Order Traversal.cc
// 8ms, 57.76%; 13.8MB, 98.59%
#include <bits/stdc++.h>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode *root) {
        vector<vector<int>> res;
        queue<TreeNode *> q;
        if (root) q.push(root);
        while (!q.empty()) {
            res.push_back(vector<int>());
            for (int i = q.size(); i > 0; --i) {
                auto tmp = q.front();
                q.pop();
                res.back().push_back(tmp->val);
                if (tmp->left) q.push(tmp->left);
                if (tmp->right) q.push(tmp->right);
            }
        }
        return res;
    }
};
